x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/PruneRedundantOrderBy.java

```diff
@@ -58,25 +58,32 @@ protected LogicalPlan rule(LogicalPlan plan) {
      * breadth-first recursion to find redundant SORTs in the children tree.
      * Returns an identity set (we need to compare and prune the exact instances)
      */
-    private Set<OrderBy> findRedundantSort(LogicalPlan plan) {
-        Set<OrderBy> result = Collections.newSetFromMap(new IdentityHashMap<>());
+    private Set<LogicalPlan> findRedundantSort(LogicalPlan plan) {
+        Set<LogicalPlan> result = Collections.newSetFromMap(new IdentityHashMap<>());
 
         Deque<LogicalPlan> toCheck = new ArrayDeque<>();
-        toCheck.push(plan);
+        toCheck.push(((UnaryPlan) plan).child());
 
-        while (true) {
-            if (toCheck.isEmpty()) {
-                return result;
-            }
+        while (toCheck.isEmpty() == false) {
             LogicalPlan p = toCheck.pop();
             if (p instanceof OrderBy ob) {
                 result.add(ob);
                 toCheck.push(ob.child());
+            } else if (p instanceof TopN childTopN && plan instanceof TopN parentTopN) {
+                // Check if a child TopN is redundant compared to a parent TopN.
+                // A child TopN is redundant if it has the same sort order as the parent.
+                // We do not need to compare their values because `PushDownAndCombineLimits`
+                // has already pushed down the lower limit value
+                if (childTopN.order().equals(parentTopN.order())) {
+                    result.add(childTopN);
+                    toCheck.push(childTopN.child());
+                }
             } else if (p instanceof SortAgnostic) {
```

Comment:
The new version of the rule is incorrect, I'm afraid.

We continue going down the plan as long as the children are `SortAgnostic`. However, being `SortAgnostic` doesn't automatically imply that the plans behave the same when we feed them an unlimited number of rows.

For instance, `INLINE STATS` is `SortAgnostic` (resp. `InlineJoin` is). But removing a `TopN` from before an `INLINE STATS` changes the meaning of the query.

Also, queries can become much more expensive: For instance, `MvExpand` is `SortAgnostic`, but `SORT a | LIMIT 10 | MV_EXPAND b | SORT a | LIMIT 10` is generally cheaper to run than just `MV_EXPAND b | SORT a | LIMIT 3`.